src/runtime/Py.cs

```diff
@@ -27,18 +27,29 @@ public static PyModule CreateScope(string name)
 
     public class GILState : IDisposable
     {
+        private static readonly ThreadLocal<int> internalGILCount = new();
```

Comment:
You can check if this resolves the performance issue with the script I published above.

This also leaves the scenario, where upon the initialization of runtime `BeginAllowThreads` was never called. In that case you don't need to explicitly acquire GIL unless you reenter .NET code from Python. E.g. in the sample benchmark with `dynamic counter` I posted above you don't actually need to call `Py.GIL` unless you called `PythonEngine.BeginAllowThreads`.